src/wallet/rpcdump.cpp

```diff
@@ -858,32 +853,32 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Incompatibility found between watchonly and keys");
         }
 
-        // Internal + Label
+        // Internal addresses should not have a label
         if (internal && data.exists("label")) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Incompatibility found between internal and label");
         }
 
-        // Keys / PubKeys size check.
-        if (!isP2SH && (keys.size() > 1 || pubKeys.size() > 1)) { // Address / scriptPubKey
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "More than private key given for one address");
-        }
-
-        // Invalid P2SH redeemScript
-        if (isP2SH && !IsHex(strRedeemScript)) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid redeem script");
+        // Force users to provide the witness script in its field rather than redeemscript
+        if (!strRedeemScript.empty() && script.IsPayToWitnessScriptHash()) {
```

Comment:
Instead of checking that `strRedeemScript` is not empty, this should check that `strWitnessScript` is empty. Or it could check both: `!strRedeemScript.empty() && strWitnessScript.empty()`.

Reply:
It should be possible to provide a P2WSH scriptPubKey without providing the witness script for it, e.g. watch only, so I don't think that's right

Reply:
I think checking both covers that case.

Reply:
Checking both would mean that if witnessScript and redeemScript were both present, it would allow it, which makes no sense if its not a P2WSH address

Reply:
@Sjors Yes, of course, just `importaddress` it. It probably works fine without this PR even.

Reply:
So we currently support importing a `3...` address as watch-only, but the wallet doesn't know or care if it's SegWit or not. The new code here is for a scenario where you do know it's SegWit `script.IsPayToWitnessScriptHash()`, which means you know the `strRedeemScript`, but you don't know the `strWitnessScript`. That seems a strange situation. Why would anyone give you the legacy redeem script but not the Segwit witness script?

Reply:
I'm very confused by your response.

I just suggested having a test that doesn't do anything but import a bech32 address without revealing the pubkey, and see if it's treated as watch-only.

Reply:
@Sjors I have no idea what you're talking about. This is about the case where you're importing a bech32 P2WSH address (no P2SH), without revealing the script inside - which is a totally reasonable thing if you don't care about solvability.

Reply:
@sipa I see, I got confused because the deleted line `if (isP2SH`.

Reply:
@sipa I see, I may have gotten confused because Github is showing `if (isP2SH && !IsHex(strRedeemScript` as the deleted line, but this thread is about an earlier code branch.
<img width="591" alt="schermafbeelding 2018-10-22 om 14 16 57" src="https://user-images.githubusercontent.com/10217/47281424-45db9880-d605-11e8-977f-3330bbd7da82.png">